File: source/F#/src/perms.fs
// Signed Permutation Cycle Counting

open System
open System.Linq
open System.Collections.Generic

let version = 8

// Select Tarjan or Loops
module Sets = Tarjan

// Generate initial permutation for each possible prefix
let enumPrefixes n k =
   let rest = List.init n id

   let rec pick k prefix rest =
      if k = 0 then
         [ prefix @ rest ]
      else
         [ for x in rest do
              let sans_x = List.filter ((<>) x) rest
              yield! pick (k - 1) (x :: prefix) sans_x ]

   pick k [] rest

// Heap's algorithm: tally all perms with a fixed length k prefix
let tallyPerms perm k work =
   let n = Array.length perm
   let swap = Parallel.swap

   let rec generate j =
      if j < k then
         work perm
      else
         generate (j - 1)

         for i = k to j - 1 do
            if (j - k) % 2 = 0 then swap perm k j else swap perm i j
            generate (j - 1)

   generate (n - 1)

// Count cycles in a signed permutation
let countCycles n (perm : int array) signs rel =
   Sets.reset rel (2 * n)

   for i = 0 to n - 1 do
      let j = perm.[i]

      if (signs >>> i) &&& 1 = 1 then
         Sets.unite rel i (j + n)
         Sets.unite rel (i + n) j
      else
         Sets.unite rel i j
         Sets.unite rel (i + n) (j + n)

   Sets.setCount rel

// Tally cycle counts across all signs for one perm
let tallySigns n (perm : int[]) (tally : int64[]) rel =
   let maxBits = 1 <<< n

   for signs = 0 to maxBits - 1 do
      let cycles = countCycles n perm signs rel
      let index = 2 * n - cycles
      tally.[index] <- tally.[index] + 1L

// Process one parcel: tally all cycle counts with given prefix
let runPrefix n k perm =
   let perm = Array.ofList perm
   let tally : int64 array = Array.zeroCreate (2 * n)
   let rel = Sets.create (2 * n)
   let work perm = tallySigns n perm tally rel

   tallyPerms perm k work
   tally

// Entry point: distribute work parcels and combine results
let runParcels n k cores =
   let zero = Array.zeroCreate (2 * n)

   enumPrefixes n k
   |> Parallel.map cores zero (runPrefix n k)
   |> function
      | [] -> zero
      | head :: tail -> List.fold (Array.map2 (+)) head tail

// Parse command-line arguments
let parseArgs argv =
   match argv with
   | [| nStr ; prefixStr ; coresStr |] ->
      match
         Int32.TryParse (nStr : string),
         Int32.TryParse (prefixStr : string),
         Int32.TryParse (coresStr : string)
      with
      | (true, n), (true, prefix), (true, cores) ->
         Some (n, prefix, cores)
      | _ ->
         eprintfn "Error: All arguments must be valid integers."
         None
   | _ ->
      eprintfn "Error: Required arguments: n prefix cores"
      None

// Entry point
[<EntryPoint>]
let main argv =
   match parseArgs argv with
   | Some (n, k, cores) ->
      printfn
         "%s v%d, n = %d, prefix = %d, cores = %d"
         Sets.name
         version
         n
         k
         cores

      let result = runParcels n k cores
      Answers.check n result
      0
   | None -> 1

File: source/haskell/src-old/Perms.hs
-- Signed Permutation Cycle Counting

import Control.Applicative ((<$>), (<*>))
import Control.Monad
import Control.Monad.ST
import Data.List (foldl')
import Data.Vector.Unboxed (Vector)
import qualified Data.Vector.Unboxed as V
import qualified Data.Vector.Unboxed.Mutable as MV
import System.Environment (getArgs)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Text.Printf (printf)
import Text.Read (readMaybe)

import Answers (check)
import Parallel (parallelMap)
import Worker (runPrefix, algName)

version :: Int
version = 17

-- Generate initial permutation for each possible prefix
enumPrefixes :: Int -> Int -> [[Int]]
enumPrefixes _ 0 = [[]]
enumPrefixes n k =
   [ x : p
   | p <- enumPrefixes n (k - 1)
   , x <- [0 .. n - 1]
   , x `notElem` p
   ]

-- Distribute work parcels and combine results
runParcels :: Int -> Int -> Vector Int
runParcels n prefix =
   let parcels = enumPrefixes n prefix
       worker p = runST (runPrefix n p)
       results = parallelMap worker parcels
       zero = V.replicate (2 * n) 0
    in foldl' (V.zipWith (+)) zero results

-- Parse command-line arguments
parseArgs :: [String] -> Maybe (Int, Int, Int)
parseArgs [n, prefix, cores] =
   (,,)
      <$> readMaybe n
      <*> readMaybe prefix
      <*> readMaybe cores
parseArgs _ = Nothing

-- Main entry point
main :: IO ()
main = do
   args <- getArgs
   case parseArgs args of
      Just (n, prefix, cores) -> do
         printf
            "%s v%d, n = %d, prefix = %d, cores = %d\n"
            algName
            version
            n
            prefix
            cores

         let result = runParcels n prefix
         let list = V.toList result

         check n list
      Nothing ->
         hPutStrLn stderr "Required arguments: n prefix cores"
            >> exitFailure
